include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -1051,6 +1054,54 @@ struct ArenaCfg : Base<OrtArenaCfg> {
 // Custom OPs (only needed to implement custom OPs)
 //
 
+/// <summary>
+/// This struct wraps a raw pointer const OrtNodeArg* that is obtained from
+/// a KernelInfo object. Use it to access the name or type info of an input
+/// or output.
+/// </summary>
+struct NodeArg {
```

Comment:
Will update to use new Ort::Base<> impl if https://github.com/microsoft/onnxruntime/pull/13215 is merged first

Reply:
This exposes the graph. Need to carefully consider the implications

Reply:
Yes, this is one of my concerns. I saw that the python bindings also expose NodeArg (albeit in a more limited fashion) so I felt compelled to try it this way to see what you'd think. https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/onnxruntime_pybind_state.cc#L1365

If we don't feel good about exposing NodeArg, then one alternative is to create more apis on top of KernelInfo to accomplish the same tasks. EX:
- KernelInfo_GetInputName(..., size_t index, ...)
- KernelInfo_GetOutputName(..., size_t index, ..)
- KernelInfo_GetInputTypeInfo(..., size_t, index, ..)
- KernelInfo_GetOutputTypeInfo(..., size_t, index, ..)

The drawback is that this requires more APIs and that we have to keep passing in the index.

Another alternative is to allocate an entirely new object that only has the data that we need. We don't have to return a pointer to an internal onnxruntime::NodeArg. I think I may like this more.